iofog-agent-daemon/src/main/java/org/eclipse/iofog/pruning/DockerPruningManager.java

```diff
@@ -91,6 +98,73 @@ public void start() throws Exception {
         }
     };
 
+    /**
+     * Gets list of unwanted docker images to be removed
+     * @return list
+     */
+    public Set<String> getUnwantedImagesList() {
+        Set<String> imageIDsToBePruned = new HashSet<>();
+        List<Image> images = docker.getImages();
+        LoggingService.logDebug(MODULE_NAME, "Total number of images already downloaded in the machine : " + images.size());
+        List<Container> nonIoFogContainers = docker.getRunningNonIofogContainers();
+        LoggingService.logDebug(MODULE_NAME, "Total number of running non iofog containers : " + nonIoFogContainers.size());
+
+        // Removes the non-ioFog running container from the images to be prune list
+        for (Container container : nonIoFogContainers) {
+            for (Image image: images) {
+                if (container.getImageId().equals(image.getId())) {
+                    LoggingService.logDebug(MODULE_NAME, "Non ioFog image not to be pruned : " + container.getImage());
+                    images.remove(image);
+                    break;
+                }
+
+            }
+        }
+        LoggingService.logDebug(MODULE_NAME, "Total number of ioFog images  : " + images.size());
+        List<Microservice> microservices = microserviceManager.getLatestMicroservices();
+        LoggingService.logDebug(MODULE_NAME, "Total number of running microservices : " + microservices.size());
+
+        // Removes the ioFog running containers from the images to be prune list
+        for (Image image: images){
+            for (Microservice microservice: microservices){
+                if (image.getRepoTags()[0] != null) {
+                    if(microservice.getImageName().equals(image.getRepoTags()[0])){
+                        if(imageIDsToBePruned.contains(image.getId())) {
+                            imageIDsToBePruned.remove(image.getId());
+                        }
+                        break;
+                    } else {
+                        imageIDsToBePruned.add(image.getId());
+                    }
+                } else {
+                    imageIDsToBePruned.add(image.getId());
+                    break;
+                }
+            }
+        }
+        LoggingService.logDebug(MODULE_NAME, "Total number of unwanted images to be pruned : " + imageIDsToBePruned.size());
+        return imageIDsToBePruned;
+    }
+
+    /**
+     * Remove unwanted docker image
+     * @param imageIDsToBePruned
+     */
+    private void removeImagesById(Set<String> imageIDsToBePruned){
+        LoggingService.logDebug(MODULE_NAME, "Start removing image by ID");
+        try {
+            for (String id: imageIDsToBePruned) {
+                LoggingService.logInfo(MODULE_NAME, "Removing unwanted image id : " + id);
+                docker.removeImageById(id);
```

Comment:
I would catch exceptions on `removeImageById` and let the loop try deleting rest of the list.

Reply:
updated